src/django_pg_migration_tools/operations.py

```diff
@@ -525,3 +583,256 @@ def describe(self) -> str:
+        if is_not_null and (not constraint_exists):
+            return
+
+        if not constraint_exists:
+            self._alter_table_not_valid_constraint(
+                schema_editor, table_name, column_name, constraint_name
+            )
+            self._validate_constraint(schema_editor, table_name, constraint_name)
+            self._alter_table_not_null(schema_editor, table_name, column_name)
+            self._alter_table_drop_constraint(
+                schema_editor, table_name, constraint_name
+            )
+            return
+        elif self._is_constraint_valid(schema_editor, constraint_name):
+            self._alter_table_not_null(schema_editor, table_name, column_name)
+            self._alter_table_drop_constraint(
+                schema_editor, table_name, constraint_name
+            )
+            return
+        else:
+            # Constraint exists and is NOT VALID.
+            self._validate_constraint(schema_editor, table_name, constraint_name)
+            self._alter_table_not_null(schema_editor, table_name, column_name)
+            self._alter_table_drop_constraint(
+                schema_editor, table_name, constraint_name
+            )
+            return
+
+    def _alter_table_not_valid_constraint(
+        self,
+        schema_editor: base_schema.BaseDatabaseSchemaEditor,
+        table_name: str,
+        column_name: str,
+        constraint_name: str,
+    ) -> None:
+        cursor = schema_editor.connection.cursor()
+        cursor.execute(
+            psycopg_sql.SQL(
+                ConstraintQueries.ALTER_TABLE_CONSTRAINT_NOT_NULL_NOT_VALID
+            ).format(
+                table_name=psycopg_sql.Identifier(table_name),
+                column_name=psycopg_sql.Identifier(column_name),
+                constraint_name=psycopg_sql.Identifier(constraint_name),
+            )
+        )
+
+    def _is_not_null(
+        self,
+        schema_editor: base_schema.BaseDatabaseSchemaEditor,
+        table_name: str,
+        column_name: str,
+    ) -> bool:
+        cursor = schema_editor.connection.cursor()
+        cursor.execute(
+            NullabilityQueries.IS_COLUMN_NOT_NULL,
+            {"table_name": table_name, "column_name": column_name},
+        )
+        return bool(cursor.fetchone())
+
+    def _get_constraint_name(self, table_name: str, column_name: str) -> str:
+        """
+        We need a unique name for the constraint.
+        We don't care too much about what the name itself turns out to be. This
+        constraint will be deleted at the end of the process anyway.
+        Here we just give it some resemblance to the table and column names the
+        constraint was created from.
+        """
+        suffix: str = django_backends_utils.names_digest(  # type:ignore[attr-defined]
+            table_name, column_name, length=10
+        )
+        return f"{table_name[:10]}_{column_name[:10]}_{suffix}"
```

Comment:
❓ In the scenario that there is a not null constraint that already exists with a different name (i.e. was generated from something else), would the creation of a second null constraint (because this would only search for this specific constraint name) covering the same column cause any issues?

Reply:
That's a great question. Having an existing constraint wouldn't cause any issues here. You can have duplicate NOT NULL constraints on the same field as long as they have different names.

However, it is possible to re-use an existing constraint that a user created for whatever reason (maybe they even forgot about it).

But I don't think it should be our responsibility to clean up the schema for the user when the action was not caused by us (unless it blocks us in some way).

At the minimum, doing that would involve adding extra introspective queries to cover these corner-cases, thus penalising regular users that won't likely have an existing constraint against that field.